### ISSUE TITLE
subset,ctd-method() inserts NA values for short elements

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#2250]
This came up in thinking about #2249; a reprex is below.  I'll need to look in the code to see what to do.  It might complicated, with ripple effects.  But I don't this will come up in the normal way of working.  This all comes down to some ERDAPP files putting single values into the variable list, instead of putting them in the attribute list.

``` r
library(oce)
#> Loading required package: gsw
data(ctd)
ctd@data$longitude <- ctd@metadata$longitude
ctd@data$latitude <- ctd@metadata$latitude
str(ctd@data)
#> List of 9
#>  $ scan       : int [1:181] 130 131 132 133 134 135 136 137 138 139 ...
#>  $ timeS      : num [1:181] 129 130 131 132 133 134 135 136 137 138 ...
#>  $ pressure   : num [1:181] 1.48 1.67 2.05 2.24 2.62 ...
#>  $ depth      : num [1:181] 1.47 1.66 2.04 2.23 2.6 ...
#>  $ temperature: num [1:181] 14.2 14.2 14.2 14.2 14.2 ...
#>  $ salinity   : num [1:181] 29.9 29.9 29.9 29.9 29.9 ...
#>  $ flag       : num [1:181] 0 0 0 0 0 0 0 0 0 0 ...
#>  $ longitude  : num -63.6
#>  $ latitude   : num 44.7
CTD <- subset(ctd, pressure < 10)
str(CTD@data)
#> List of 9
#>  $ scan       : int [1:36] 130 131 132 133 134 135 136 137 138 139 ...
#>  $ timeS      : num [1:36] 129 130 131 132 133 134 135 136 137 138 ...
#>  $ pressure   : num [1:36] 1.48 1.67 2.05 2.24 2.62 ...
#>  $ depth      : num [1:36] 1.47 1.66 2.04 2.23 2.6 ...
#>  $ temperature: num [1:36] 14.2 14.2 14.2 14.2 14.2 ...
#>  $ salinity   : num [1:36] 29.9 29.9 29.9 29.9 29.9 ...
#>  $ flag       : num [1:36] 0 0 0 0 0 0 0 0 0 0 ...
#>  $ longitude  : num [1:36] -63.6 NA NA NA NA ...
#>  $ latitude   : num [1:36] 44.7 NA NA NA NA ...
```

<sup>Created on 2024-09-22 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by dankelley** [comment]
I've altered `subset,ctd-method()` so that it now only does the subset operation on items in the `@data` slot if the subset logical vector is of the same length as the data item.  (Man, that's a confusing sentence.)

My reprex now looks as follows.  I will not be pushing to GH until maybe this afternoon, though, because I'll want to do more testing. 

``` r
library(oce)
#> Loading required package: gsw
data(ctd)
ctd@data$longitude <- ctd@metadata$longitude
ctd@data$latitude <- ctd@metadata$latitude
str(ctd@data)
#> List of 9
#>  $ scan       : int [1:181] 130 131 132 133 134 135 136 137 138 139 ...
#>  $ timeS      : num [1:181] 129 130 131 132 133 134 135 136 137 138 ...
#>  $ pressure   : num [1:181] 1.48 1.67 2.05 2.24 2.62 ...
#>  $ depth      : num [1:181] 1.47 1.66 2.04 2.23 2.6 ...
#>  $ temperature: num [1:181] 14.2 14.2 14.2 14.2 14.2 ...
#>  $ salinity   : num [1:181] 29.9 29.9 29.9 29.9 29.9 ...
#>  $ flag       : num [1:181] 0 0 0 0 0 0 0 0 0 0 ...
#>  $ longitude  : num -63.6
#>  $ latitude   : num 44.7
CTD <- subset(ctd, pressure < 10)
str(CTD@data)
#> List of 9
#>  $ scan       : int [1:36] 130 131 132 133 134 135 136 137 138 139 ...
#>  $ timeS      : num [1:36] 129 130 131 132 133 134 135 136 137 138 ...
#>  $ pressure   : num [1:36] 1.48 1.67 2.05 2.24 2.62 ...
#>  $ depth      : num [1:36] 1.47 1.66 2.04 2.23 2.6 ...
#>  $ temperature: num [1:36] 14.2 14.2 14.2 14.2 14.2 ...
#>  $ salinity   : num [1:36] 29.9 29.9 29.9 29.9 29.9 ...
#>  $ flag       : num [1:36] 0 0 0 0 0 0 0 0 0 0 ...
#>  $ longitude  : num -63.6
#>  $ latitude   : num 44.7
```

<sup>Created on 2024-09-22 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by dankelley** [comment]
I've added the following to the test suite.
```R
# https://github.com/dankelley/oce/issues/2250
test_that("subset handles uncommon-length items", {
    data(ctd)
    A <- ctd
    As <- subset(A, pressure < 5)
    B <- ctd
    B@data$longitude <- B@metadata$longitude
    B@data$latitude <- B@metadata$latitude
    Bs <- subset(B, pressure < 5)
    expect_equal(1L, length(Bs@data$longitude))
    expect_equal(1L, length(Bs@data$latitude))
    for (name in names(A@data)) {
        expect_equal(As@data[[name]], Bs@data[[name]])
    }
})
```

**Comment by dankelley** [comment]
This is fixed in commit 309aba6aba8e934e6a08aee7fa8b192cb84a60b0 of the "develop" branch.  (The new test code will ensure that this behaviour remains fixed.). Since this has built and tested both locally and on the 5 machines used by R-CMD-check, I'm closing the issue.
